src/crh/sha512/r1cs_utils.rs

```diff
@@ -0,0 +1,122 @@
+use crate::Vec;
+
+use ark_ff::PrimeField;
+use ark_r1cs_std::bits::{boolean::Boolean, uint64::UInt64, uint8::UInt8, ToBitsGadget};
+use ark_relations::r1cs::SynthesisError;
+use core::iter;
+
+/// Extra traits not automatically implemented by UInt64
```

Comment:
We don't need this in two different locations, right?

Reply:
We don't. I wasn't sure how you would like to resolve it, so I went with code duplication and a note in the pull request :)
I think we could create e.g. a subdirectory for SHA and use a macro to generate these traits for both 512-byte and 256-byte algorithms.